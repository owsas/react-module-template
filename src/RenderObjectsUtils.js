import CurrencyFormatter from 'currency-formatter';

import { Constants as constants } from '@owsas/geopromos-web-constants';

/**
 * Helps getting the information to be shown
 * on Cards and other UI items for the classes
 * of the Database
 */
export default class RenderObjectsUtils {
  /**
   * Gets the info to be shown for the promo
   * @param {Parse.Object} obj
   */
  static getInfoForPromo(obj) {
    return {
      img: (obj.get('pics') && obj.get('pics')[0]) || obj.get('picurl') || constants.img.default,
      title: obj.get('name'),
      description: obj.get('description'),
      date: obj.createdAt,
      badges: obj.get('trends'),
      to: `/promos/${obj.id}`,
      author: obj.get('brand') && obj.get('brand').get('name'),
    };
  }

  /**
   * Gets the info to be shown for the product
   * @param {Parse.Object} obj
   */
  static getInfoForProduct(obj) {
    return {
      img: (obj.get('pics') && obj.get('pics')[0]) || constants.img.default,
      title: obj.get('defaultName'),
      description: obj.get('description'),
      price: obj.get('price') ? CurrencyFormatter.format(
        obj.get('price'),
        { code: (obj.get('currencyCode') || 'COP') },
      ) : undefined,
      badges: obj.get('trends'),
      to: `/products/${obj.id}`,
      author: obj.get('brand') && obj.get('brand').get('name'),
    };
  }

  /**
   * Gets the info to be shown for the brand
   * @param {Parse.Object} obj
   */
  static getInfoForBrand(obj) {
    return {
      img: obj.get('picUrl') ||
        (obj.get('picture') && obj.get('picture').url()) ||
        constants.img.default,
      title: obj.get('name'),
      description: obj.get('description'),
      to: `/brands/${obj.id}`,
    };
  }

  /**
   * Gets the info to be shown for the category
   * @param {Parse.Object} obj
   */
  static getInfoForCategory(obj) {
    return {
      img: (obj.get('logo') && obj.get('logo').url()) ||
        constants.img.default,
      title: obj.get('defaultName'),
      alias: obj.get('alias'),
      to: `/search?c=${obj.get('alias')}`,
    };
  }

  /**
   * Gets the info to be shown for the subcategory
   * @param {Parse.Object} obj
   */
  static getInfoForSubcategory(obj) {
    return {
      img: (obj.get('logo') && obj.get('logo').url()) ||
        constants.img.default,
      title: obj.get('defaultName'),
      alias: obj.get('alias'),
      to: `/search?s=${obj.get('alias')}`,
    };
  }


  /**
   * Gets the info to be shown for the given object
   * @param {Parse.Object} obj
   */
  static getInfoForObject(obj) {
    let params;

    switch (obj.className) {
      case 'Promo':
        params = RenderObjectsUtils.getInfoForPromo(obj);
        break;
      case 'Product':
        params = RenderObjectsUtils.getInfoForProduct(obj);
        break;
      case 'Brand':
        params = RenderObjectsUtils.getInfoForBrand(obj);
        break;
      case 'Category':
        params = RenderObjectsUtils.getInfoForCategory(obj);
        break;
      case 'Subcategory':
        params = RenderObjectsUtils.getInfoForSubcategory(obj);
        break;
      default:
        params = {
          title: obj.get('name') || obj.get('title'),
          to: '/404',
        };
    }

    return params;
  }
}
