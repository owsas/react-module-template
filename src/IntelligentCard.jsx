/* eslint-disable no-multiple-empty-lines  */

import React from 'react';
import PropTypes from 'prop-types';
import Parse from 'parse';
import { DataBrowser as Data } from '@owsas/geopromos-private-api/out/DataBrowser';

import { Card } from '@owsas/geopromos-web-css';
import RenderObjectsUtils from './RenderObjectsUtils';


// Renders a card in an intelligent way,
// looking for the className of the Parse.Object
// to show, and displaying the right information
const IntelligentCard = (props) => {
  const params = RenderObjectsUtils.getInfoForObject(props.obj);
  const date = params.date && Data.getDateSinceString(params.date);

  return (
    <div style={{ display: 'block', textDecoration: 'none' }} onClick={props.onClick}>
      <Card img={params.img}>
        {/* The title of the card */}
        {params.title && <p className={`${props.ellipsis && 'ellipsis'} cardTitle`}><b>{params.title}</b></p>}

        {/* If extended, show the author */}
        {props.extended && params.author && <p className="cardAuthor">{params.author}</p>}

        {/* The price is always shown */}
        {params.price &&
        <p style={{ color: 'green' }} className="cardPrice"><b>{params.price}</b></p>
          }

        {/* The price is always shown */}
        {date && <p className="date cardDate"> <i className="fa fa-clock-o" /> {date}</p>}
      </Card>
    </div>
  );
};


// eslint-disable-next-line
IntelligentCard.defaultProps = {
  ellipsis: false,
  extended: false,
  onClick: undefined,
};

IntelligentCard.propTypes = {
  obj: PropTypes.instanceOf(Parse.Object).isRequired,
  ellipsis: PropTypes.bool,
  extended: PropTypes.bool, // shows more information
  onClick: PropTypes.func,
};

export default IntelligentCard;
