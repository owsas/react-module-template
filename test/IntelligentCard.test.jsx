import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Parse from 'parse/node';

import IntelligentCard from '../src/IntelligentCard';
import RenderObjectsUtils from '../src/RenderObjectsUtils';

const brand = new Parse.Object('Brand');
brand.set('name', 'La marca');
brand.set('description', 'Muy importante');
brand.set('picUrl', 'def://123');
brand.id = 'abc123';

const obj = new Parse.Object('Promo');
obj.set('pics', ['abc://123']);
obj.set('name', 'La promo');
obj.set('trends', ['#123']);
obj.set('description', 'La descripciÃ³n de la promo');
obj.set('brand', brand);
obj.id = '123';

describe('Rendered View', () => {
  const m = shallow(<IntelligentCard obj={obj} extended />);

  const res = RenderObjectsUtils.getInfoForObject(obj);

  describe('Card information', () => {
    const card = m.at(0);

    test('.cardTitle should have the correct title', () => {
      expect(card.find('.cardTitle').text()).toEqual(res.title);
    });

    test('.cardAuthor should have the correct author', () => {
      expect(card.find('.cardAuthor').text()).toEqual(res.author);
    });

    test('.cardPrice should have the correct price', () => {
      const prod = new Parse.Object('Product');
      prod.set('price', 12312);
      prod.set('currencyCode', 'CO');

      const m2 = shallow(<IntelligentCard obj={prod} extended />);

      const res2 = RenderObjectsUtils.getInfoForObject(prod);

      expect(m2.find('.cardPrice').text()).toEqual(res2.price);
    });
  });


  test('should match snapshot', () => {
    expect(toJSON(m)).toMatchSnapshot();
  });
});
