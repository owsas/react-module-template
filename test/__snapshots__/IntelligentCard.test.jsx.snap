// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendered View should match snapshot 1`] = `
<div
  style={
    Object {
      "display": "block",
      "textDecoration": "none",
    }
  }
>
  <l
    contentStyle={Object {}}
    img="abc://123"
    imgStyle={Object {}}
  >
    <p
      className="false cardTitle"
    >
      <b>
        La promo
      </b>
    </p>
    <p
      className="cardAuthor"
    >
      La marca
    </p>
  </l>
</div>
`;
